### PR TITLE
Non-record: Random Linear Maps + Learned Adapters (val_bpb=1.93 (1.607 outside 10mins), 2.4MB artifact (1.9mb outside 10mins)

### PR DESCRIPTION
**Pull request by fielding**
# Random Linear Maps + Learned Adapters

**val_bpb: 1.93** (10 min, 8xH100 SXM) | **1.607** (3.75hr, 4xH200) | **2.4 MB artifact**

## The Idea

What if 90% of your model's weights were just noise? 

Each linear layer gets a random base weight matrix. They are generated from a fixed seed at init time, like `42` or `1337`. Those base weights cost zero bytes in the artifact because they're regenerated from the seed at eval. Only small rank-16 adapters (LoRA-style A and B matrices) are learned and stored. Think of it like giving someone a house made of random LEGO bricks and a small bag of correct ones... they have to figure out which random bricks are useful and nudge the rest into place with the adapters.

A 512-dim, 5-layer model normally stores around 25M parameters. This approach stores 2.2M. The other 90% are deterministic noise from a seed. The artifact is ~2.4 MB (10 min, 8xH100) or 1.9 MB (3.75hr long run), well under the 16 MB budget.

### Depth Sweep (fixed 40-min training budget, rank=16, 512-dim)

| Layers | Step time | Steps      | Float BPB | Int6 BPB  | Artifact   |
|--------|-----------|------------|-----------|-----------|------------|
| 3      | 43ms      | 50,000     | 1.756     | 1.720     | 1.5 MB     |
| **4**  | **55ms**  | **43,879** | **1.648** | **1.665** | **1.7 MB** |
| **5**  | **67ms**  | **35,618** | **1.638** | **1.656** | **1.9 MB** |
| 6      | 85ms      | 28,255     | 1.654     | 1.675     | 2.1 MB     |
| 8      | 109ms     | 22,031     | 1.651     | 1.673     | 2.5 MB     |
| 11     | 142ms     | 16,922     | 1.770     | 1.794     | 3.1 MB     |
| 20     | 254ms     | 9,457      | 3.551     | 3.552     | 4.3 MB     |

4-5 layers is the sweet spot. Shallower models train faster and rack up more steps, which matters more than depth when the base weights carry no learned information. Go too shallow (3L) and there isn't enough compositional depth for language. Go too deep (20L) and gradients straight up can't propagate through that many random projections. The model with 20 layers learned nothing.

### Rank Sweep (768-dim, 11L, fixed 40-min budget)

| Rank   | Stored Params | BPB      | Artifact   |
|--------|---------------|----------|------------|
| **16** | **3.2M**      | **2.77** | **4.2 MB** |
| 32     | 5.2M          | 3.20     | 5.7 MB     |
| 64     | 9.3M          | 3.55     | 9.5 MB     |

This one's counterintuitive. Smaller adapters win. Rank 16 crushes rank 32 and 64 because the larger adapters need more training steps to converge, and the fixed time budget punishes them for it. This sweep was run at 768-dim/11L, a harder setting than the depth sweep. The directional finding (smaller rank wins) should hold at 512-dim, but the absolute BPB numbers aren't comparable to the depth table.

### Scaling with Training Time (5L, rank=16, 512-dim)

| Steps       | Training Time | Float BPB | Sliding BPB | Artifact   |
|-------------|---------------|-----------|-------------|------------|
| 4,212       | 10 min        | 2.58      | —           | 1.9 MB     |
| 20,000      | 49 min        | 1.81      | —           | 1.9 MB     |
| 50,000      | 2 hr          | 1.64      | 1.63        | 1.9 MB     |
| **200,000** | **3.75 hr**   | **1.66**  | **1.607**   | **1.9 MB** |

Sliding BPB keeps improving with more steps, though with diminishing returns. The float BPB at 200K (1.66) is slightly worse than 50K (1.64), likely from training instability mid-run (loss spiked around step 104K) that the warmdown didn't fully recover from. Sliding window evaluation smooths this out, which is why the sliding BPB still improved. The model hasn't fully converged.

## Architecture

```python
class RandomLinearWithAdapter(nn.Module):
    def __init__(self, in_features, out_features, seed, rank=16):
        # Random base: NOT stored, regenerated from seed
        g = torch.Generator().manual_seed(seed)
        self.register_buffer('base_weight',
            torch.randn(out_f, in_f, generator=g) / math.sqrt(in_f),
            persistent=False)

        # Learned adapter: stored in artifact
        self.adapter_A = nn.Parameter(torch.randn(rank, in_f) * 0.01)
        self.adapter_B = nn.Parameter(torch.zeros(out_f, rank))

    def forward(self, x):
        W = self.base_weight + self.adapter_B @ self.adapter_A
        return F.linear(x, W, self.bias)
```

`persistent=False` means the base weight never hits the state_dict. At load time, `__init__` regenerates it from the seed. Each layer gets a unique seed from its index.

Every attention projection (Q, K, V, output) and MLP layer (fc, proj) uses RandomLinearWithAdapter. Embeddings, norms, and other small parameters are fully learned.

### Model Configuration

| Component     | Detail                             |
|---------------|------------------------------------|
| Layers        | 5                                  |
| Dim           | 512                                |
| Heads         | 8 (4 KV, GQA)                      |
| MLP           | 3x, relu-squared                   |
| Adapter rank  | 16                                 |
| Stored params | 2.2M (11% of full model)           |
| Random params | ~20M (89%, not stored)             |
| EMA           | 0.997                              |
| Training      | Muon + AdamW, 200K steps on 4xH200 |

## What I Found

**Depth has a sweet spot with random projections.** 4-5 layers wins at a fixed time budget. More depth means fewer training steps, and step count is king when base weights carry no information. 20 layers learned nothing... literally.

**Smaller adapters optimize better.** Rank 16 beats 32 and 64. There's a capacity-optimization tradeoff here... bigger adapters have more capacity but need more steps to figure out how to use it.

**Random projections can do language modeling.** A ~2.4 MB model with 90% random weights hits 1.93 BPB in 10 minutes on 8xH100 (1.607 with extended training). The naive baseline (fully learned, 13.5 MB) hits 1.224 BPB. The gap is real, but the fact that it works at all is the interesting part. A natural follow-up is comparing against a size-matched fully learned model at ~2.4 MB to isolate the contribution of random maps vs model capacity. That experiment is planned but not yet run.

**The artifact is hilariously small.** ~2.4 MB is 15% of the 16 MB budget. You could fit six of these models in one artifact. Ensembles, multi-model voting, whatever you want... there's room.

## 3-Seed Validation (8xH100 SXM, 600s)

| Seed | Sliding BPB | Artifact |
|------|-------------|----------|
| 42   | 1.9313      | 2.4 MB   |
| 1337 | 1.9079      | 2.4 MB   |
| 2024 | 1.9439      | 2.4 MB   |

Seeds used for random base weight generation are derived from SEED + layer_index. They were chosen arbitrarily, not searched.

## Run Commands

```bash
# 40-min depth sweep (run for NUM_LAYERS=3,4,5,6,8,11,20)
RANDOM_LINEAR=1 ADAPTER_RANK=16 MODEL_DIM=512 NUM_HEADS=8 \
  NUM_KV_HEADS=4 NUM_LAYERS=5 EMA_ENABLED=1 \
  ITERATIONS=50000 MAX_WALLCLOCK_SECONDS=2400 \
  torchrun --standalone --nproc_per_node=4 train_gpt.py

# Long run (best config)
RANDOM_LINEAR=1 ADAPTER_RANK=16 MODEL_DIM=512 NUM_HEADS=8 \
  NUM_KV_HEADS=4 NUM_LAYERS=5 EMA_ENABLED=1 \
  ITERATIONS=200000 MAX_WALLCLOCK_SECONDS=14400 \
  torchrun --standalone --nproc_per_node=4 train_gpt.py
```

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Non-record: Random Linear Maps + Learned Adapters (val_bpb=1.93, 2.4MB artifact)

**BPB:** 1.93 | **Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

**What I found in the code** (head SHA `6b211a8bc44f`, file `records/track_non_record_16mb/2026-03-26_RandomLinearMaps/train_gpt.py`):

Static code review found no TTT adaptation function, no SLOT optimization loop, no n-gram-cache class, and no pre-quant val-token fine-tune. The eval path uses the standard sliding-window stride-64 pattern. The submission is a pure-neural architecture iteration on the standard SP1024/SP4096/SP8192 baseline.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 0.13s, dim=512, layers=9, vocab=1024, code=83802 B, SMOKE_TEST_PASS

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the classification pass — this looks like a clean pure-neural iteration on the standard baseline.

*Auto-classification caveat:* this review was drafted by the AST-based classifier. If there's a non-standard eval mechanism (logit postprocessing, hedge mixing, etc.) that I missed because it's factored into a helper file or a non-standard function name, please flag it and I'll re-run the audit manually.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 0.13s, dim=512, layers=9, vocab=1024, code=83802 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*

**Comment by fielding** [comment]
## Followup: tried MoE on top of RLM, didn't beat it.

I hinted in toward the end of this PR about the next steps involving MoE or some ensemble. The pitch is structurally clean: RLM made each linear cheap (frozen random base + low-rank adapter), so adding experts costs basically just the adapter, not the full weight. The random base is "free"... one int32 seed regenerates it at load time. So you get many experts for cheap.

That story holds, but in practice...  

**Hardware note**: RunPod could only spare 4xH100 SXM at the time (we'd have grabbed 8 if available). The ablation runs are 20 min on 4xH100, which is roughly the same total compute as the record-track 10 min on 8xH100 budget. The 60-min headline run is just "we had the credits, why not."  

Each block's MLP becomes N experts on the input projection (per-expert seeded random base + adapter), one shared output projection, learned router with a depth-aware bias that prefers expert L % E at init. Top-2 dense dispatch (sparse breaks torch.compile(fullgraph=True) because of nonzero()). Switch-Transformer aux loss.  

**result**: val_bpb 2.29 with E=4, 60 min on 4xH100. RLM baseline above is 1.93. So 0.36 BPB worse on the same architecture skeleton.  

Expert-count sweep at fixed 20-min wallclock on 4xH100, seq=1024:  

| E | val_bpb | Steps |
|---|---------|-------|
| 2 | **2.69** | 2263 (got 40% warmdown) |
| 4 | 2.84 | 1526 |
| 6 | **2.78** | 1170 (best non-warmdown) |
| 8 | 2.88 | 949 |
| 10 | 3.17 | 806 |
| 12 | 2.94 | 698 |
| 14 | 2.93 | 608 |
| 16 | 3.55 | 546 (bootstrap never broke) |

E=2 wins because it's the only one that reaches warmdown. Among non-warmdown runs, E=6 is the sweet spot... capacity helps, but only up to a point before per-step cost eats your training budget.

**Why it doesn't pay off:** dense routing means MLP compute scales with E, so step count drops. At a fixed wallclock, fewer steps is the bigger problem than fewer parameters. Warmdown is also doing more work than I'd guessed... the runs that reach it leapfrog runs that don't, regardless of expert count. Same lesson as my distillation followup honestly. The architectural overhead has to clear a real bar to beat just training the simpler thing for longer.  

Not opening a separate PR for it... the result is too far behind to be worth the review noise. Code lives on a local branch in case anyone wants to poke at it.
